File: app/internal/publisher/template/java/HardwareDetector.java
package com.websmith.sdk;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.InetAddress;
import java.net.NetworkInterface;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Enumeration;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class HardwareDetector {

    private String fingerprint;
    private Map<String, String> identifiers;

    public String getFingerprint() {
        if (fingerprint == null) {
            Map<String, String> ids = collectIdentifiers();
            String combined = buildCombinedString(ids);
            fingerprint = hashIdentifiers(combined);
            identifiers = ids;
        }
        return fingerprint;
    }

    public Map<String, String> getIdentifiers() {
        if (identifiers == null) {
            getFingerprint();
        }
        return identifiers != null ? identifiers : new HashMap<>();
    }

    private Map<String, String> collectIdentifiers() {
        Map<String, String> ids = new HashMap<>();
        String cpuId = getCpuId();
        if (cpuId != null) {
            ids.put("cpu_id", cpuId);
        }
        String motherboardId = getMotherboardId();
        if (motherboardId != null) {
            ids.put("motherboard_id", motherboardId);
        }
        if (!ids.containsKey("motherboard_id")) {
            String networkId = getNetworkId();
            if (networkId != null) {
                ids.put("network_id", networkId);
            }
        }
        String osInfo = getOsInfo();
        if (osInfo != null) {
            ids.put("os_info", osInfo);
        }
        return ids;
    }

    private String getCpuId() {
        String os = System.getProperty("os.name", "").toLowerCase();
        try {
            if (os.contains("win")) {
                return getCpuIdWindows();
            } else if (os.contains("mac")) {
                return getCpuIdDarwin();
            } else if (os.contains("nix") || os.contains("nux")) {
                return getCpuIdLinux();
            }
        } catch (Exception ignored) {}
        return System.getProperty("os.arch");
    }

    private String getCpuIdWindows() {
        try {
            Process process = Runtime.getRuntime().exec(
                new String[]{"wmic", "cpu", "get", "ProcessorId", "/value"});
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream()));
            String line;
            while ((line = reader.readLine()) != null) {
                if (line.startsWith("ProcessorId=")) {
                    String id = line.substring("ProcessorId=".length()).trim();
                    if (!id.isEmpty()) return id;
                }
            }
        } catch (Exception ignored) {}
        return System.getProperty("os.arch");
    }

    private String getCpuIdDarwin() {
        try {
            Process process = Runtime.getRuntime().exec(
                new String[]{"sysctl", "-n", "hw.model"});
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream()));
            String model = reader.readLine();
            if (model != null && !model.trim().isEmpty()) {
                return "mac-" + model.trim();
            }
        } catch (Exception ignored) {}
        try {
            Process process = Runtime.getRuntime().exec(
                new String[]{"sysctl", "-n", "machdep.cpu.brand_string"});
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream()));
            String brand = reader.readLine();
            if (brand != null && !brand.trim().isEmpty()) {
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                byte[] hash = md.digest(brand.trim().getBytes(StandardCharsets.UTF_8));
                return bytesToHex(hash).substring(0, 16);
            }
        } catch (Exception ignored) {}
        return System.getProperty("os.arch");
    }

    private String getCpuIdLinux() {
        try {
            Process process = Runtime.getRuntime().exec(
                new String[]{"cat", "/proc/cpuinfo"});
            BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream()));
            String line;
            String serial = null;
            String vendor = "";
            String family = "";
            while ((line = reader.readLine()) != null) {
                if (line.toLowerCase().startsWith("serial")) {
                    String[] parts = line.split(":");
                    if (parts.length > 1) {
                        serial = parts[1].trim();
                    }
                } else if (line.startsWith("vendor_id")) {
                    String[] parts = line.split(":");
                    if (parts.length > 1) vendor = parts[1].trim();
                } else if (line.startsWith("cpu family")) {
                    String[] parts = line.split(":");
                    if (parts.length > 1) family = parts[1].trim();
                }
            }
            if (serial != null && !serial.isEmpty()) {
                return "cpu-" + serial;
            }
            if (!vendor.isEmpty() && !family.isEmpty()) {
                return vendor + "-" + family;
            }
        } catch (Exception ignored) {}
        return System.getProperty("os.arch");
    }

    private String getMotherboardId() {
        String os = System.getProperty("os.name", "").toLowerCase();
        try {
            if (os.contains("win")) {
                Process process = Runtime.getRuntime().exec(
                    new String[]{"wmic", "baseboard", "get", "SerialNumber", "/value"});
                BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()));
                String line;
                while ((line = reader.readLine()) != null) {
                    if (line.startsWith("SerialNumber=")) {
                        String serial = line.substring("SerialNumber=".length()).trim();
                        if (!serial.isEmpty() && !"To be filled by O.E.M.".equalsIgnoreCase(serial)
                            && !"Default string".equalsIgnoreCase(serial)) {
                            return "mb-" + serial;
                        }
                    }
                }
            } else if (os.contains("nix") || os.contains("nux")) {
                Process process = Runtime.getRuntime().exec(
                    new String[]{"dmidecode", "-s", "baseboard-serial-number"});
                BufferedReader reader = new BufferedReader(
                    new InputStreamReader(process.getInputStream()));
                String serial = reader.readLine();
                if (serial != null) {
                    serial = serial.trim();
                    if (!serial.isEmpty() && !"To be filled by O.E.M.".equalsIgnoreCase(serial)
                        && !"Default string".equalsIgnoreCase(serial)) {
                        return "mb-" + serial;
                    }
                }
            }
        } catch (Exception ignored) {}
        return null;
    }

    private String getNetworkId() {
        try {
            List<String> macs = new ArrayList<>();
            Enumeration<NetworkInterface> interfaces = NetworkInterface.getNetworkInterfaces();
            while (interfaces.hasMoreElements()) {
                NetworkInterface ni = interfaces.nextElement();
                if (ni.isLoopback() || ni.isVirtual()) continue;
                byte[] mac = ni.getHardwareAddress();
                if (mac != null) {
                    StringBuilder sb = new StringBuilder();
                    for (byte b : mac) {
                        sb.append(String.format("%02x:", b));
                    }
                    if (sb.length() > 0) sb.deleteCharAt(sb.length() - 1);
                    macs.add(sb.toString());
                }
            }
            if (!macs.isEmpty()) {
                MessageDigest md = MessageDigest.getInstance("SHA-256");
                String combined = "net-" + String.join(":", macs.subList(0, Math.min(3, macs.size())));
                byte[] hash = md.digest(combined.getBytes(StandardCharsets.UTF_8));
                return bytesToHex(hash).substring(0, 16);
            }
        } catch (Exception ignored) {}
        return null;
    }

    private String getOsInfo() {
        return System.getProperty("os.name", "") + "-" + System.getProperty("os.version", "");
    }

    private String buildCombinedString(Map<String, String> identifiers) {
        StringBuilder sb = new StringBuilder();
        String[] keys = {"cpu_id", "motherboard_id", "network_id"};
        for (int i = 0; i < keys.length; i++) {
            if (identifiers.containsKey(keys[i])) {
                if (sb.length() > 0) sb.append("|");
                sb.append(identifiers.get(keys[i]));
            }
        }
        return sb.toString();
    }

    private String hashIdentifiers(String data) {
        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(data.getBytes(StandardCharsets.UTF_8));
            return bytesToHex(hash);
        } catch (Exception e) {
            return "fallback-" + UUID.randomUUID().toString();
        }
    }

    private String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}

File: app/internal/publisher/template/dotnet/HardwareDetector.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net.NetworkInformation;
using System.Runtime.InteropServices;
using System.Security.Cryptography;
using System.Text;

namespace WebsmithSDK;

public class HardwareDetector
{
    private string? _fingerprint;
    private Dictionary<string, string>? _identifiers;

    public string GetFingerprint()
    {
        if (_fingerprint == null)
        {
            var ids = CollectIdentifiers();
            var combined = BuildCombinedString(ids);
            _fingerprint = HashIdentifiers(combined);
            _identifiers = ids;
        }
        return _fingerprint;
    }

    public Dictionary<string, string> GetIdentifiers()
    {
        if (_identifiers == null) GetFingerprint();
        return _identifiers ?? new Dictionary<string, string>();
    }

    private Dictionary<string, string> CollectIdentifiers()
    {
        var ids = new Dictionary<string, string>();

        var cpuId = GetCpuId();
        if (cpuId != null) ids["cpu_id"] = cpuId;

        var motherboardId = GetMotherboardId();
        if (motherboardId != null) ids["motherboard_id"] = motherboardId;

        if (!ids.ContainsKey("motherboard_id"))
        {
            var networkId = GetNetworkId();
            if (networkId != null) ids["network_id"] = networkId;
        }

        var osInfo = GetOsInfo();
        if (osInfo != null) ids["os_info"] = osInfo;

        return ids;
    }

    private string? GetCpuId()
    {
        try
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return GetCpuIdWindows();
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return GetCpuIdDarwin();
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
                return GetCpuIdLinux();
        }
        catch { }
        return RuntimeInformation.OSArchitecture.ToString();
    }

    private string? GetCpuIdWindows()
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "wmic",
                Arguments = "cpu get ProcessorId /value",
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            using var proc = Process.Start(psi);
            if (proc != null)
            {
                var output = proc.StandardOutput.ReadToEnd();
                proc.WaitForExit(2000);
                foreach (var line in output.Split('\n', '\r'))
                {
                    if (line.StartsWith("ProcessorId="))
                    {
                        var id = line["ProcessorId=".Length..].Trim();
                        if (!string.IsNullOrEmpty(id)) return id;
                    }
                }
            }
        }
        catch { }
        return RuntimeInformation.OSArchitecture.ToString();
    }

    private string? GetCpuIdDarwin()
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "sysctl",
                Arguments = "-n hw.model",
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            using var proc = Process.Start(psi);
            if (proc != null)
            {
                var model = proc.StandardOutput.ReadToEnd().Trim();
                proc.WaitForExit(2000);
                if (!string.IsNullOrEmpty(model)) return "mac-" + model;
            }
        }
        catch { }
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "sysctl",
                Arguments = "-n machdep.cpu.brand_string",
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true
            };
            using var proc = Process.Start(psi);
            if (proc != null)
            {
                var brand = proc.StandardOutput.ReadToEnd().Trim();
                proc.WaitForExit(2000);
                if (!string.IsNullOrEmpty(brand))
                {
                    return Convert.ToHexStringLower(
                        SHA256.HashData(Encoding.UTF8.GetBytes(brand)))[..16];
                }
            }
        }
        catch { }
        return RuntimeInformation.OSArchitecture.ToString();
    }

    private string? GetCpuIdLinux()
    {
        try
        {
            var cpuinfo = File.ReadAllText("/proc/cpuinfo");
            foreach (var line in cpuinfo.Split('\n'))
            {
                if (line.StartsWith("Serial", StringComparison.OrdinalIgnoreCase))
                {
                    var parts = line.Split(':');
                    if (parts.Length > 1)
                    {
                        var serial = parts[1].Trim();
                        if (!string.IsNullOrEmpty(serial)) return "cpu-" + serial;
                    }
                }
            }
            string vendor = "", family = "";
            foreach (var line in cpuinfo.Split('\n'))
            {
                if (line.StartsWith("vendor_id"))
                {
                    var parts = line.Split(':');
                    if (parts.Length > 1) vendor = parts[1].Trim();
                }
                else if (line.StartsWith("cpu family"))
                {
                    var parts = line.Split(':');
                    if (parts.Length > 1) family = parts[1].Trim();
                }
            }
            if (!string.IsNullOrEmpty(vendor) && !string.IsNullOrEmpty(family))
                return $"{vendor}-{family}";
        }
        catch { }
        return RuntimeInformation.OSArchitecture.ToString();
    }

    private string? GetMotherboardId()
    {
        try
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                var psi = new ProcessStartInfo
                {
                    FileName = "wmic",
                    Arguments = "baseboard get SerialNumber /value",
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                };
                using var proc = Process.Start(psi);
                if (proc != null)
                {
                    var output = proc.StandardOutput.ReadToEnd();
                    proc.WaitForExit(2000);
                    foreach (var line in output.Split('\n', '\r'))
                    {
                        if (line.StartsWith("SerialNumber="))
                        {
                            var serial = line["SerialNumber=".Length..].Trim();
                            if (!string.IsNullOrEmpty(serial)
                                && !serial.Equals("To be filled by O.E.M.", StringComparison.OrdinalIgnoreCase)
                                && !serial.Equals("Default string", StringComparison.OrdinalIgnoreCase))
                                return "mb-" + serial;
                        }
                    }
                }
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                var psi = new ProcessStartInfo
                {
                    FileName = "dmidecode",
                    Arguments = "-s baseboard-serial-number",
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                };
                using var proc = Process.Start(psi);
                if (proc != null)
                {
                    var serial = proc.StandardOutput.ReadToEnd().Trim();
                    proc.WaitForExit(2000);
                    if (!string.IsNullOrEmpty(serial)
                        && !serial.Equals("To be filled by O.E.M.", StringComparison.OrdinalIgnoreCase)
                        && !serial.Equals("Default string", StringComparison.OrdinalIgnoreCase))
                        return "mb-" + serial;
                }
            }
        }
        catch { }
        return null;
    }

    private string? GetNetworkId()
    {
        try
        {
            var macs = new List<string>();
            foreach (var ni in NetworkInterface.GetAllNetworkInterfaces())
            {
                if (ni.OperationalStatus == OperationalStatus.Up
                    && ni.NetworkInterfaceType != NetworkInterfaceType.Loopback)
                {
                    var mac = ni.GetPhysicalAddress().ToString();
                    if (mac.Length > 0)
                    {
                        macs.Add(string.Join(":",
                            Enumerable.Range(0, 6).Select(i => mac.Substring(i * 2, 2))));
                    }
                }
            }
            if (macs.Count > 0)
            {
                var combined = "net-" + string.Join(":", macs.Take(3));
                return Convert.ToHexStringLower(
                    SHA256.HashData(Encoding.UTF8.GetBytes(combined)))[..16];
            }
        }
        catch { }
        return null;
    }

    private string? GetOsInfo()
    {
        return RuntimeInformation.OSDescription;
    }

    private static string BuildCombinedString(Dictionary<string, string> identifiers)
    {
        var parts = new List<string>();
        foreach (var key in new[] { "cpu_id", "motherboard_id", "network_id" })
        {
            if (identifiers.TryGetValue(key, out var value))
                parts.Add(value);
        }
        return string.Join("|", parts);
    }

    private static string HashIdentifiers(string data)
    {
        return Convert.ToHexStringLower(
            SHA256.HashData(Encoding.UTF8.GetBytes(data)));
    }
}
